### PR TITLE
AP_HAL_SITL: init ArduSub pwm_output to 1500

### PR DESCRIPTION
**Pull request by clydemcqueen**
Problem:

When using SIM_JSON with ArduSub the first ~500 packets sent to ArduPilotPlugin have pwm values [1000, ...]. This sends the sub flying for a few seconds.

It appears that `SITL_State::_simulator_servos` initializes pwm_output to [1000, ...] and these values are sent to ArduPilotPlugin by SIM_JSON before any ArduSub code is called.

Proposed solution:

Add code to `SITL_State::_simulator_servos` to initialize pwm_output to [1500, ...] if the vehicle is ArduSub. 

Tests:

* Tested with Ignition Gazebo, ardupilot_gazebo and a BlueROV2 model, see https://github.com/clydemcqueen/bluerov2_ignition/tree/clyde_relay
* `Tools/scripts/build_all.sh` builds everything cleanly

### CODE REVIEW

**Comment by clydemcqueen** [comment]
FYI @Williangalvani @srmainwaring

**Review by khancyr — COMMENTED** [review]
The issue is the change here will impact all sub model and setting 1500 as default output will now break the default SITL model.

1500 should be use for reversible motors like on rover.

I am on phone but I am remembering that I have already solve this issue for rover or gazebo sim in the past and we should have a flags somewhere to prevent sending the value to the backend until the controller really output something

**Comment by Williangalvani** [comment]
> 1500 should be use for reversible motors like on rover.

Sub uses reversible thrusters like rover, where 1500 means they are stopped. Have I missed something?

That said, at least the commit text should be prepended with `AP_HAL_SITL:`

File: libraries/AP_HAL_SITL/SITL_State.cpp [hunk 1]
```diff
@@ -698,6 +698,10 @@ void SITL_State::_simulator_servos(struct sitl_input &input)
         if (_vehicle == Rover) {
             pwm_output[0] = pwm_output[1] = pwm_output[2] = pwm_output[3] = 1500;
         }
+        if (_vehicle == ArduSub) {
+            pwm_output[0] = pwm_output[1] = pwm_output[2] = pwm_output[3] =
+                    pwm_output[4] = pwm_output[5] = pwm_output[6] = pwm_output[7] = 1500;
```
Comment (ES-Alexander) [inline libraries/AP_HAL_SITL/SITL_State.cpp:703]: This seems potentially problematic, given it's operating on the outputs directly. Since sub vehicles can have fewer than 8 thrusters this could cause unexpected issues with other peripherals in the additional slots (e.g. lights, gripper, etc). If possible it would perhaps be best to turn off the outputs (set them to 0 maybe?) and have no output PWM until the controller starts up and starts commanding intentional output values.

If turning off the outputs is impossible then I suppose this is at least preferable to the current behaviour of setting definitely incorrect values for the thrusters, but it's still not desirable behaviour for a vehicle (just a bit less undesirable).

Comment (clydemcqueen) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: Thanks for the review, @ES-Alexander 

The challenge here is the boot sequence for SITL with --model=JSON. Basically, the SITL startup code happens very early, and the 1st thing SIM::JSON tries to do is establish a connection to the FDM (e.g., Gazebo via ardupilot_gazebo). The protocol requires that SITL send a packet with servo data, then it stalls -- blocking the boot -- waiting for a response. Once the connection is established SITL must continue to send packets with servo data to keep the connection alive.

Right now the packets are initialized with [1000, ...], and ardupilot_gazebo doesn't know that these are bad values, so until ArduPilot finishes booting the thrusters are just going crazy.

I think the real fix is probably to modify the protocol between SIM::JSON and the FDM to mark the servo data as "not initialized." But that require changes to all SIM::JSON clients.

This fix is much simpler, and it works for the 2 basic SITL sub cases (6 or 8 thrusters). It also seems to be the same way that copter does it, so at least the code is localized.

Comment (ES-Alexander) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: > The protocol requires that SITL send a packet with servo data, then it stalls -- blocking the boot -- waiting for a response. Once the connection is established SITL must continue to send packets with servo data to keep the connection alive.

Does "servo data" require valid servo values (e.g. 1000-2000), or is it possible to specify 0 as each value and effectively command "turn off the outputs"? If all zeros isn't possible can it be 1500 for the first output (which is most likely a thruster) and then set zero for the rest?

If the first 8 outputs all require valid servo-range PWM values then I agree that your proposed solution is probably the best we can do for now, but if it's possible to turn off all or most of the outputs instead then that would seem to be preferable given it avoids potentially 'causing damage' to/with non-thruster peripherals that may be connected to the main outputs.

Comment (clydemcqueen) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: > Does "servo data" require valid servo values (e.g. 1000-2000), or is it possible to specify 0 as each value and effectively command "turn off the outputs"? If all zeros isn't possible can it be 1500 for the first output (which is most likely a thruster) and then set zero for the rest?

Relevant ardupilot_gazebo code: https://github.com/ArduPilot/ardupilot_gazebo/blob/main/src/ArduPilotPlugin.cc#L1173

My example sdf file: https://github.com/clydemcqueen/bluerov2_ignition/blob/main/models/bluerov2/model.sdf#L458

TL;DR ardupilot_gazebo treats all inputs as valid, including 0.

I don't know of any other clients, but presumably if they exist might do something else.

I should also point out that SITL currently supports just 2 frames: vectored and vectored_6dof. Folks trying to simulate other frames will have to modify the simulation code.

Comment (srmainwaring) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: @clydemcqueen it would be reasonable to interpret pwm = 0 as 'no command' in ardupilot_gazebo. I don't think there are any cases in the example models where this would cause valid input data ranges to fail. We'd need a patch to check for this and skip sending commands on to the motors, and could instead mark the model as initialising (as opposed to no connection).

Comment (clydemcqueen) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: @srmainwaring thanks for the suggestion. I coded up a quick version of this idea in ardupilot and ardupilot_gazebo and ran some tests and it works well.

On the ardupilot_gazebo side there is a bit more state to manage, but it is pretty simple.

On the ardupilot side we can implement this by removing the code in `SITL_State::_simulator_servos()` that writes to `pwm_output`, and rely on the fact that the compiler already initializes `pwm_output` to [0, ...]. We can add a comment to describe the situation for future maintainers. This is clean, but it does affect all simulations (models, vehicles and frames.) Thoughts?

Comment (ES-Alexander) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: I'm for it, but also definitely biased since I don't work with the other vehicle types so don't know how it may or may not affect them. That said, my personal opinion is that sending default valid actuator outputs on startup _without_ the input of an actual controller is a poor way of managing things, especially if it's effectively just being used as a "do nothing" workaround to communicate that the vehicle is in fact connected - this revised approach seems (to me at least) likely to be an improvement for everyone :-)

Comment (Williangalvani) [reply libraries/AP_HAL_SITL/SITL_State.cpp:703]: I'll add a these to next week's devcall so we can discuss if this would be a better approach

**Review by Williangalvani — APPROVED** [review]

**Comment by tridge** [comment]
needs rebase

**Comment by clydemcqueen** [comment]
Rebased. Thanks @tridge and @Williangalvani
